### PR TITLE
Add support for DeepSpeed sequence parallelism (Ulysses)

### PR DESCRIPTION
**Pull request by ronald-d-rogers**
# What does this PR do?

Adds support for sequence parallel to Tranformers.

**❗ The sister PR to `accelerate` (https://github.com/huggingface/accelerate/pull/3299) must be merged first ❗**

The PR mostly pulls in and updates changes made in @zeyugao's very thorough PR here:
https://github.com/huggingface/accelerate/pull/2877

And pulls in @samedjacob's PR here:
https://github.com/huggingface/transformers/pull/32305

To support changes made by @samejacobs in DeepSpeed, to make sequence parallel integration with HF cleaner here:
https://github.com/microsoft/DeepSpeed/pull/5774

And to respond to comments to the his PR to `transformers` here #32305.

Note that my testing has revealed that all of @zeyugao's changes are indeed required for this integration to work.

I've added two decorators:
- `deepspeed_ulysses_attention`: Instead of modifying the global `_flash_attention_forward` we wrap it instead as suggested by @ArthurZucker.
- `support_deepspeed_ulysses`: An extension of the concept above to add support for sequence parallel to modules by injecting required variables like `sp_group_size`.

I am not married to any of these changes, and am completely open to suggestions.

The only major addition made in my PR is to find the right place to shard the inputs for sequence parallelism to work.

I added a method to `Trainer` called `_finalize_inputs` which is meant to run right before the inputs are passed to the model. I toyed around with sharding the sequences in various parts of the data loading pipeline (see commits if you are curious), but ultimately determined that it had to be done in trainers, right before the inputs are passed to the model, otherwise libraries with custom trainers (i.e. `trl`) would have to refactor the way their trainers prepare data.

As things currently stand, **the sister PR to `accelerate` (https://github.com/huggingface/accelerate/pull/3299) must be merged first**, so that the new flags added to `HFDeepSpeedConfig` for sequence parallelism are available, but if desired we can likely make these PRs orthogonal by caching the same flags in `transformers` for now. 

So far everything appears to work (forward/backward pass tested on 4xA10s and 8xA100s). I will be doing more testing soon (i.e. testing loss is same, tuning a model & evaluating, etc.).

~~Note that we should also update `loss_utils.py`'s `fixed_cross_entropy` method to use 
[`vocab_sequence_parallel_cross_entropy`](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/sequence/cross_entropy.py#L59) when sequence parallel is enabled, but this method currently does not allow you to propagate the args `ignore_index` and `reduction`. I will create a PR for this in DeepSpeed shortly.~~

I have also integrated DeepSpeed's Ulysses loss ([`vocab_sequence_parallel_cross_entropy`](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/sequence/cross_entropy.py#L59)) into `loss_utils.py` and my tests are currently showing good loss curves.

I am also beginning to think we should use the `DistributedAttention` that is in `DeepSpeed` here instead of decorating the global flash attention function, similar to the way @zeyugao had it, as it seems to have a lot of functionality:
https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/sequence/layer.py#L300

I will likely test this soon as well.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @sunmarc @muellerzr

**Comment by SunMarc** [comment]
cc @XuehaiPan

**Comment by ronald-d-rogers** [comment]
Hello all, just to provide a progress update, I am currently tuning and evaluating a few models to validate that this is all working correctly.

One thing I have noticed is that the loss for sequence parallel seems to be slightly different, but in the same ballpark. I am still trying to figure out exactly why or if this matters. There seems to be a single token which is being clipped and I have yet to determine what that token is and how to preserve it.

Also, though this works with TRL's SFT without issue the loss for DPO seems to currently be way too high. I am trying to investigate what the cause of the is.

**Comment by ronald-d-rogers** [comment]
Hello all, I was busy with something else and just got back to work on this.

I have tuned and evaluated several Llama 3.1 8B models on several different tasks (classification/summarization) and am currently comparing the results of all of the runs.

So far I am definitely noticing some oddities between DeepSpeed ZeRO-3 and DeepSpeed Ulysses.

The `vocab_sequence_parallel_cross_entropy` as it is currently implemented/integrated doesn't appear to work at all. Hugging Face's default loss aggregation across GPUs seems to work for classification but I am currently running to issues with summarization. I am currently tuning a few more models to see if the issues goes away (with more epochs), but if not I think I will have to start digging into how to loss is implemented.

Currently from what I can tell, we aren't doing `softmax` across the whole sequence in calculating the cross-entropy loss, which I think could be causing issues, but am not sure.

Either way, I will post charts and updates within a day or two.

**Comment by ronald-d-rogers** [comment]
Hello all, I have been getting pulled away from this but it have been slowing working on it on my 20% time. I'm happy to report that after much trial and error I think I have everything working as expected! I'm going to retune all models and report back soon. It seems like the issue was with the last logit being lost in each shard during token shifting to align preds w/ labels. I will retune all models and report back.

**Comment by Falko1** [comment]
Hi, 
thanks for the great work! I tried to get your half-finished PR to work but failed. Instead, I got this (https://github.com/princeton-nlp/ProLong) Sequence Parallel implementation to work, which integrates Ulysses-style SP into the HF trainer. Maybe you can get some inspiration there to finish the PR. 
I'm glad to try it once it is merged!

**Comment by ronald-d-rogers** [comment]
@Falko1: Thank you very much for this link. Yes, I am completely stuck trying to get other integrations to work. Maybe I will pivot to this instead.

**Review by ArthurZucker — COMMENTED** [review]
Hey! BTW I think you need a small rebase, we modified the attention format quite a lot!

**Comment by ArthurZucker** [comment]
We are looking into RingAttention as well could be good!

**Comment by ronald-d-rogers** [comment]
@ArthurZucker: [Picotron](https://github.com/huggingface/picotron)?

**Comment by ArthurZucker** [comment]
Yeah, with the attention refactor we think things should be easier now!
